test/systemtest/utils.py

```diff
@@ -226,6 +243,38 @@ def close(self):
 
         super().close()
 
+    def log_add_marker(self, text: str):
+        """ Write a marker text into the UART send and receive log files """
+        text_b = text.encode('UTF-8')
+        self._log_to_device_fp.write(text_b)
+        self._log_from_device_fp.write(text_b)
+
+    def drain_in(self, timeout=None, silence_time=.5):
+        """ Read all available input data
+
+        Args:
+            timeout: Maximum time this function blocks, in seconds.
+            silence_time: Consider the input drained if no new data can be read
+                after this time, in seconds.
+        Returns:
+            The data read from the device.
+        """
+        if timeout != None:
+            t_end = time.time() + timeout
+
+        read_data = b''
+
+        first_iteration = False
+        while first_iteration or self.in_waiting:
+            if timeout != None and time.time() >= t_end:
+                break
+
+            read_data += self.read(self.in_waiting)
+            first_iteration = False
+            time.sleep(silence_time)
```

Comment:
I think I'm probably just missing something, but how does this loop exit early when we have no output for `silence_time`, but haven't yet hit `timeout`?

Reply:
Ah, right, thanks.